File: Liftr/ExternalRouteDedupPolicy.swift
import Foundation
import HealthKit

enum ExternalRouteDedupPolicy {
    static let liftrRouteProviderKey = "com.lilru.liftr.external_route_provider"
    static let liftrRouteIdKey = "com.lilru.liftr.external_route_id"
    static let liftrHasCustomRouteKey = "com.lilru.liftr.has_custom_route"

    static func shouldSkip(
        workout: HKWorkout,
        existingRoutes: [HKWorkoutRoute],
        provider: String,
        providerActivityId: String
    ) -> Bool {
        if existingRoutes.contains(where: { route in
            (route.metadata?[liftrRouteIdKey] as? String) == providerActivityId
        }) {
            return true
        }
        if workout.metadata?[liftrHasCustomRouteKey] as? Bool == true {
            return true
        }
        if !existingRoutes.isEmpty {
            return true
        }
        return false
    }

    static func routeMetadata(provider: String, providerActivityId: String) -> [String: Any] {
        [
            liftrHasCustomRouteKey: true,
            liftrRouteProviderKey: provider,
            liftrRouteIdKey: providerActivityId,
        ]
    }
}

File: android/app/src/main/java/com/lilru/liftr/externalroute/ExternalRouteDedupPolicy.kt
package com.lilru.liftr.externalroute

object ExternalRouteDedupPolicy {
    const val LIFTR_ROUTE_PROVIDER_KEY = "com.lilru.liftr.external_route_provider"
    const val LIFTR_ROUTE_ID_KEY = "com.lilru.liftr.external_route_id"
    const val LIFTR_HAS_CUSTOM_ROUTE_KEY = "com.lilru.liftr.has_custom_route"

    fun routeMetadata(provider: String, providerActivityId: String): Map<String, String> =
        mapOf(
            LIFTR_HAS_CUSTOM_ROUTE_KEY to "true",
            LIFTR_ROUTE_PROVIDER_KEY to provider,
            LIFTR_ROUTE_ID_KEY to providerActivityId
        )
}
